kube-runtime/src/controller/runner.rs

```diff
@@ -332,4 +373,82 @@ mod tests {
             Error::Readiness(delayed_init::InitDropped)
         ));
     }
+
+    // A Future that is Ready after the specified duration from its initialization.
+    struct DurationalFuture {
+        start: Instant,
+        ready_after: Duration,
+    }
+
+    impl DurationalFuture {
+        fn new(expires_in: Duration) -> Self {
+            let start = Instant::now();
+            DurationalFuture {
+                start,
+                ready_after: expires_in,
+            }
+        }
+    }
+
+    impl Future for DurationalFuture {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let now = Instant::now();
+            if now.duration_since(self.start) > self.ready_after {
+                Poll::Ready(())
+            } else {
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn runner_should_respect_max_concurrent_executions() {
```

Comment:
does the runner respect it if the messages are `Eq`? e.g. does it interact with deduplication?

Reply:
the check for concurrency happens before the deduplication logic, so no it shouldn't interfere with deduplication. so if a message is already present in `slots` and while the runner is at its limit, the same message arrives then it'll be marked as pending without even any attempt to deduplicate